### PR TITLE
Improvements to Galaxy and Pulsar job execution

### PR DESCRIPTION
**Pull request by jmchilton**
For easy merging - the combination of #7924, #8597, and #8449 - Pulsar and Galaxy job running PRs. See individual PRs for more details.

### CODE REVIEW

File: lib/galaxy/jobs/runners/__init__.py [hunk 1]
```diff
@@ -475,8 +475,12 @@ def _finish_or_resubmit_job(self, job_state, job_stdout, job_stderr, job_id=None
...
             exit_code = job_state.read_exit_code()
 
-            tool_stdout_path = os.path.join(job_wrapper.working_directory, "tool_stdout")
-            tool_stderr_path = os.path.join(job_wrapper.working_directory, "tool_stderr")
+            outputs_directory = os.path.join(job_wrapper.working_directory, "outputs")
+            if not os.path.exists(outputs_directory):
+                outputs_directory = job_wrapper.working_directory
+
+            tool_stdout_path = os.path.join(outputs_directory, "tool_stdout")
+            tool_stderr_path = os.path.join(outputs_directory, "tool_stderr")
```
Comment (bgruening) [inline lib/galaxy/jobs/runners/__init__.py:483]: Is it time to do what the TODO below says?

File: test/integration/test_containerized_jobs.py [hunk 2]
```diff
@@ -109,6 +116,10 @@ def test_build_mulled(self):
         assert any([True for d in response if d['dependency_type'] == self.container_type])
 
 
+# Singularity 2.4 in the official Vagrant issue has some problems running this test
```
Comment (bgruening) [inline test/integration/test_containerized_jobs.py:119]: @jmchilton for your information, we have generated all 32.000 Singularity images with Singularity 3.x. This means they are not compatible with Singularity 2.x.

Comment (jmchilton) [reply test/integration/test_containerized_jobs.py:119]: Good to know, thanks. I need a better test setup for Singularity for sure - luckily all the problems I landed up encountering with Pulsar actually were for all containers 😆.

File: lib/galaxy/jobs/runners/__init__.py (hunk 1 above)
Comment (jmchilton) [reply lib/galaxy/jobs/runners/__init__.py:483]: We could think about doing that this cycle, maybe not in this PR though?

Comment (bgruening) [reply lib/galaxy/jobs/runners/__init__.py:483]: Not in this PR, no. Thanks @jmchilton!

**Review by bgruening — APPROVED** [review]

**Comment by nsoranzo** [comment]
@jmchilton This PR (commit 8d78dc1cbe3ce768ca7cd3ac7c162c961efb31f0 in particular according to `git bisect`) broke uploads to data libraries from Galaxy filesystem via API, only when symlinking (i.e. `payload['link_data_only'] = 'link_to_files'`).

See https://travis-ci.org/galaxyproject/bioblend/jobs/588394255

**Pull request merged**